test/fake_cipherprov.c

```diff
@@ -0,0 +1,334 @@
+        && OSSL_PARAM_BLD_push_utf8_string(tmpl, FAKE_CIPHER_PARAM_KEY_NAME,
+                                           ctx->key_name, strlen(ctx->key_name)) == 0)
+        goto end;
+
+    if (selection & OSSL_SKEYMGMT_SELECT_SECRET_KEY
+        && OSSL_PARAM_BLD_push_octet_ptr(tmpl, OSSL_SKEY_PARAM_RAW_BYTES,
+                                         &(ctx->key), FAKE_KEY_LEN) == 0)
+        goto end;
+
+    if ((params = OSSL_PARAM_BLD_to_param(tmpl)) == NULL)
+        goto end;
+
+    ret = param_callback(params, cbarg);
+
+ end:
+    OSSL_PARAM_free(params);
+    OSSL_PARAM_BLD_free(tmpl);
+    return ret;
+}
+
+static const OSSL_DISPATCH fake_skeymgmt_funcs[] = {
+    { OSSL_FUNC_SKEYMGMT_FREE, (void (*)(void))fake_skeymgmt_free },
+    { OSSL_FUNC_SKEYMGMT_IMPORT, (void (*)(void))fake_skeymgmt_import },
+    { OSSL_FUNC_SKEYMGMT_EXPORT, (void (*)(void))fake_skeymgmt_export },
+    OSSL_DISPATCH_END
+};
+
+static const OSSL_ALGORITHM fake_skeymgmt_algs[] = {
+    { "fake_cipher", FAKE_CIPHER_FETCH_PROPS, fake_skeymgmt_funcs, "Fake Cipher Key Management" },
+    { NULL, NULL, NULL, NULL }
+};
+static OSSL_FUNC_cipher_newctx_fn fake_newctx;
+static void *fake_newctx(void *provctx)
+{
+    return OPENSSL_zalloc(sizeof(PROV_CIPHER_FAKE_CTX));
+}
+
+static OSSL_FUNC_cipher_freectx_fn fake_freectx;
+static void fake_freectx(void *vctx)
+{
+    OPENSSL_free(vctx);
+}
+
+static int fake_skey_init(PROV_CIPHER_FAKE_CTX *ctx, void *pkeyparam,
+                          const unsigned char *iv, size_t ivlen,
+                          const OSSL_PARAM params[])
+{
+    if (pkeyparam != NULL)
+        memcpy(ctx, pkeyparam, sizeof(PROV_CIPHER_FAKE_CTX));
+
+    return 1;
+}
+
+static OSSL_FUNC_cipher_encrypt_skey_init_fn fake_skey_einit;
+static int fake_skey_einit(void *vctx, void *pkeyparam,
+                           const unsigned char *iv, size_t ivlen,
+                           const OSSL_PARAM params[])
+{
+    PROV_CIPHER_FAKE_CTX *ctx = (PROV_CIPHER_FAKE_CTX *)vctx;
+
+    if (fake_skey_init(ctx, pkeyparam, iv, ivlen, params) != 1)
+        return 0;
+
+    return 1;
+}
+
+static OSSL_FUNC_cipher_decrypt_skey_init_fn fake_skey_dinit;
+static int fake_skey_dinit(void *vctx, void *pkeyparam,
+                           const unsigned char *iv, size_t ivlen,
+                           const OSSL_PARAM params[])
+{
+    PROV_CIPHER_FAKE_CTX *ctx = (PROV_CIPHER_FAKE_CTX *)vctx;
+
+    if (fake_skey_init(ctx, pkeyparam, iv, ivlen, params) != 1)
+        return 0;
+
+    return 1;
+}
+
+static OSSL_FUNC_cipher_cipher_fn fake_cipher;
+static int fake_cipher(void *vctx, unsigned char *out, size_t *outl,
+                       size_t outsize, const unsigned char *in, size_t inl)
+{
+    PROV_CIPHER_FAKE_CTX *ctx = (PROV_CIPHER_FAKE_CTX *)vctx;
+    size_t i;
+
+    if (outsize < inl)
+        return 0;
+    if (out != NULL && in != out)
+        memcpy(out, in, inl);
+    for (i = 0; i < inl; i++)
+        out[i] ^= ctx->key[i % FAKE_KEY_LEN];
```

Comment:
```suggestion
        out[i] ^= ctx->key[i % ctx->used_len];
```

Reply:
There is no used_len variable in ctx, perhaps we can use sizeof(ctx->key) ...

Reply:
Yes, sorry for my mistake.  I do think we should prefer to use the actual length of the provided key rather than including the zero padding.

All in all, not a big issue since this is part of the tests.